File: common/src/main/java/ink/anur/timewheel/TimedTask.java
package ink.anur.timewheel;

import ink.anur.io.common.ShutDownHooker;

/**
 * Created by Anur IjuoKaruKas on 2018/10/16
 * <p>
 * 需要延迟执行的任务，放在槽 {@link Bucket} 里面
 */
public class TimedTask {

    /**
     * 延迟多久执行时间
     */
    private long delayMs;

    /**
     * 过期时间戳
     */
    private long expireTimestamp;

    /**
     * 任务
     */
    private Runnable task;

    /**
     * 是否被取消
     */
    protected volatile boolean cancel;

    protected Bucket bucket;

    protected TimedTask next;

    protected TimedTask pre;

    public String desc;

    private ShutDownHooker sdh = new ShutDownHooker();

    public TimedTask(long delayMs, Runnable task) {
        this.delayMs = delayMs;
        this.task = task;
        this.bucket = null;
        this.next = null;
        this.pre = null;
        this.expireTimestamp = System.currentTimeMillis() + delayMs;
        this.cancel = false;
    }

    public void cancel() {
        sdh.shutdown();
        cancel = true;
    }

    protected void setExpireTimestamp(long t) {
        this.expireTimestamp = t;
    }

    public boolean isCancel() {
        return cancel;
    }

    public Runnable getTask() {
        return task;
    }

    public long getDelayMs() {
        return delayMs;
    }

    public long getExpireTimestamp() {
        return expireTimestamp;
    }

    public ShutDownHooker getSdh() {
        return sdh;
    }

    @Override
    public String toString() {
        return desc;
    }
}

File: common/src/main/java/ink/anur/timewheel/TimedTask.kt
package ink.anur.timewheel

import ink.anur.util.TimeUtil

/**
 * Created by Anur on 2020/7/11
 */
class TimedTask(private val delayMs: Long, private val task: Runnable) {

    /**
     * 过期时间戳
     */
    private var expireTimestamp: Long = TimeUtil.getTime() + delayMs

    /**
     * 是否被取消
     */
    @Volatile
    private var cancel = false

    var bucket: Bucket? = null

    var next: TimedTask? = null

    var pre: TimedTask? = null

    fun cancel() {
        cancel = true
    }

    fun isCancel(): Boolean {
        return cancel
    }

    fun getTask(): Runnable {
        return task
    }

    fun getExpireTimestamp(): Long {
        return expireTimestamp
    }
}